pkg/apis/networking/v1alpha1/clusteringress_types.go

```diff
@@ -0,0 +1,379 @@
+	// ClusterIngress, if left unspecified.
+	// +optional
+	Hosts []string `json:"hosts,omitempty"`
+
+	// SecretName is the name of the secret used to terminate SSL traffic.
+	SecretName string `json:"secretName,omitempty"`
+
+	// SecretNamespace is the namespace of the secret used to terminate SSL traffic.
+	SecretNamespace string `json:"secretNamespace,omitempty"`
+
+	// ServerCertificate identifies the certificate filename in the secret.
+	// Defaults to `tls.cert`.
+	// +optional
+	ServerCertificate string `json:"serverCertificate,omitempty"`
+
+	// PrivateKey identifies the private key filename in the secret.
+	// Defaults to `tls.key`.
+	// +optional
+	PrivateKey string `json:"privateKey,omitempty"`
+}
+
+// ClusterIngressRule represents the rules mapping the paths under a specified host to
+// the related backend services. Incoming requests are first evaluated for a host
+// match, then routed to the backend associated with the matching ClusterIngressRuleVlue.
+type ClusterIngressRule struct {
+	// Host is the fully qualified domain name of a network host, as defined
+	// by RFC 3986. Note the following deviations from the "host" part of the
+	// URI as defined in the RFC:
+	// 1. IPs are not allowed. Currently a ClusterIngressRuleV can only apply to the
+	//	  IP in the Spec of the parent ClusterIngress.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//	  Currently the port of an ClusterIngress is implicitly :80 for http and
+	//	  :443 for https.
+	// Both these may change in the future.
+	// Incoming requests are matched against the host before the ClusterIngressRuleValue.
+	// If the host is unspecified, the ClusterIngress routes all traffic based on the
+	// specified ClusterIngressRuleValue.
+	// +optional
+	Hosts []string `json:"hosts,omitempty"`
+
+	// ClusterIngressRuleValue represents a rule to route requests for this ClusterIngressRule.
+	// If unspecified, the rule defaults to a http catch-all. Whether that sends
+	// just traffic matching the host to the default backend or all traffic to the
+	// default backend, is left to the controller fulfilling the ClusterIngress. Http is
+	// currently the only supported ClusterIngressRuleValue.
+	//
+	// NOTE: We could inline ClusterIngressRuleValue into the parent struct, but grouping things
+	//       here to better highlight the boundary of the "OneOf" restriction.
+	// +optional
+	ClusterIngressRuleValue `json:",inline,omitempty"`
+}
+
+// ClusterIngressRuleValue represents a rule to apply against incoming requests. If the
+// rule is satisfied, the request is routed to the specified backend. Currently
+// mixing different types of rules in a single ClusterIngress is disallowed, so exactly
+// one of the following must be set.
+type ClusterIngressRuleValue struct {
+
+	// HTTP is the only supported ClusterIngressRuleValue currently so
+	// it is not optional.  However, in the future when we have others
+	// we may it will become so.  +optional
```

Comment:
Is this comment copied from Ingress? Are we considering other protocols in the foreseeable future?